frontend/src/stimulus/controllers/dynamic/work-packages/activities-tab/index.controller.ts

```diff
@@ -0,0 +1,422 @@
+import * as Turbo from '@hotwired/turbo';
+import { Controller } from '@hotwired/stimulus';
+import {
+  ICKEditorInstance,
+} from 'core-app/shared/components/editor/components/ckeditor/ckeditor.types';
+import { KeyCodes } from 'core-app/shared/helpers/keyCodes.enum';
+
+export default class IndexController extends Controller {
+  static values = {
+    updateStreamsUrl: String,
+    sorting: String,
+    pollingIntervalInMs: Number,
+    filter: String,
+    userId: Number,
+    workPackageId: Number,
+  };
+
+  static targets = ['journalsContainer', 'buttonRow', 'formRow', 'form'];
+
+  declare readonly journalsContainerTarget:HTMLElement;
+  declare readonly buttonRowTarget:HTMLInputElement;
+  declare readonly formRowTarget:HTMLElement;
+  declare readonly formTarget:HTMLFormElement;
+
+  declare updateStreamsUrlValue:string;
+  declare sortingValue:string;
+  declare lastUpdateTimestamp:string;
+  declare intervallId:number;
+  declare pollingIntervalInMsValue:number;
+  declare filterValue:string;
+  declare userIdValue:number;
+  declare workPackageIdValue:number;
+  declare localStorageKey:string;
+
+  private handleWorkPackageUpdateBound:EventListener;
+  private handleVisibilityChangeBound:EventListener;
+  private rescueEditorContentBound:EventListener;
+
+  connect() {
+    this.setLocalStorageKey();
+    this.setLastUpdateTimestamp();
+    this.setupEventListeners();
+    this.handleInitialScroll();
+    this.startPolling();
+    this.populateRescuedEditorContent();
+  }
+
+  disconnect() {
+    this.rescueEditorContent();
+    this.removeEventListeners();
+    this.stopPolling();
+  }
+
+  private setLocalStorageKey() {
+    // scoped by user id in order to avoid data leakage when a user logs out and another user logs in on the same browser
+    // TODO: when a user logs out, the data should be removed anyways in order to avoid data leakage
+    this.localStorageKey = `work-package-${this.workPackageIdValue}-rescued-editor-data-${this.userIdValue}`;
```

Comment:
we have a generalized approach for this now using the autosave feature of CKEditor. But it doesn't engage directly on beforeunload, but rather after a delay of inputting it. So it might still cause data loss if you're switching back and forth quickly. Whether it's worth to keep this in here too, I do not know

https://github.com/opf/openproject/pull/16476

Reply:
Good to see a generalized approach for recovering client side ckeditor data. I would argue that in the case of the activity tab, it is quite likely that a user (accidentally) switches a tab and would still be confused why the unsaved data is gone. Recovering the data with the approach you're mentioning is then possible but IMO not convenient enough. As a user I would much prefer not to recover my data manually. This is especially important as a tab switch with the angular implementation of the activities did not lead to a ckeditor data loss. My intention was to keep this behavior. Thoughts?